freqtrade/exchange/common.py

```diff
@@ -1,12 +1,16 @@
 import asyncio
 import logging
 import time
-from functools import wraps
+from functools import partial, wraps
 
 from freqtrade.exceptions import DDosProtection, RetryableOrderError, TemporaryError
+from freqtrade.mixins import LoggingMixin
 
 
 logger = logging.getLogger(__name__)
+logging_mixin = LoggingMixin(logger)
```

Comment:
the only thing that slightly still concerns me is this section here.

it's creating a singleton (actually 3) - which can potentially cause problems (especially to tests).
While currently, i don't think it's creating actual problems to tests - it might cause random test failures depending on the randomness used in the test.

I'm however currently lacking a better solution (in theory, LoggingMixin could be tied to the exchange instance - but that'd again assume that retrier is only ever going to be used on an exchange class.

Reply:
I don't think that it might cause problems. At the very end LoggingMixin is used "mixed" with other classes, but actually it means that there are several "instances" of this "class" which are spread around and serving the same purpose (allowing parts of the code to "log once").
In this case there isn't a real class which is mixed with LoggingMixin, but only and directly an instance of the latter, serving exactly the same purpose ("log once").

I've seen all other changes that you committed: understood and agreed.

How to continue now with this pull? Do you think that further changes are needed?

Reply:
> but actually it means that there are several "instances" of this "class" which are spread around and serving the same purpose (allowing parts of the code to "log once").

All other instances will be destroyed / recreated in the following instances:
* `/reload_config`
* new test case
* parent class is destroyed and recreated

Neither of which is true for the above instance, which is only ever created when this file is imported (which will only happen once per python process - as python caches/remembers imports).

I'll need to think and test more with this to determine if it can become a problem before deciding on a way forward.

Reply:
I think with the new approach (only create the LoggingMixing instance when necessary) we can have a good compromise.
it'll be cached for kucoin users with this problem - and will not affect any other instance (Every class requires a little bit of memory - so we can avoid that if possible).

Reply:
Indeed. That's a good compromise. Thanks!